x-pack/packages/kbn-elastic-assistant/impl/assistant/use_conversation/index.tsx

```diff
@@ -56,21 +60,78 @@ interface SetConversationProps {
 }
 
 interface UseConversation {
-  appendMessage: ({ conversationId: string, message: Message }: AppendMessageProps) => Message[];
+  appendMessage: ({ conversationId, message }: AppendMessageProps) => Message[];
+  amendMessage: ({ conversationId, content }: AmendMessageProps) => void;
   appendReplacements: ({
     conversationId,
     replacements,
   }: AppendReplacementsProps) => Record<string, string>;
   clearConversation: (conversationId: string) => void;
   createConversation: ({ conversationId, messages }: CreateConversationProps) => Conversation;
   deleteConversation: (conversationId: string) => void;
+  removeLastMessage: (conversationId: string) => Message[];
   setApiConfig: ({ conversationId, apiConfig }: SetApiConfigProps) => void;
   setConversation: ({ conversation }: SetConversationProps) => void;
 }
 
 export const useConversation = (): UseConversation => {
   const { allSystemPrompts, assistantTelemetry, setConversations } = useAssistantContext();
 
+  /**
+   * Removes the last message of conversation[] for a given conversationId
+   */
+  const removeLastMessage = useCallback(
+    (conversationId: string) => {
+      let messages: Message[] = [];
+      setConversations((prev: Record<string, Conversation>) => {
+        const prevConversation: Conversation | undefined = prev[conversationId];
+
+        if (prevConversation != null) {
+          messages = prevConversation.messages.slice(0, prevConversation.messages.length - 1);
+          const newConversation = {
+            ...prevConversation,
+            messages,
+          };
+          return {
+            ...prev,
+            [conversationId]: newConversation,
+          };
+        } else {
+          return prev;
+        }
+      });
+      return messages;
+    },
+    [setConversations]
+  );
+
+  /**
+   * Updates the last message of conversation[] for a given conversationId with provided content
+   */
+  const amendMessage = useCallback(
+    ({ conversationId, content }: AmendMessageProps) => {
+      setConversations((prev: Record<string, Conversation>) => {
+        const prevConversation: Conversation | undefined = prev[conversationId];
+
+        if (prevConversation != null) {
+          const message = prevConversation.messages.pop() as unknown as Message;
```

Comment:
Consider avoiding this combination of local mutation and casting

Reply:
++, not sure what happened to my comment here, but it's showing on the main PR description at least: https://github.com/elastic/kibana/pull/169729#discussion_r1378178537

If `messages.length === 0`, this will blow up with the spread on the next line.